compiler/noirc_evaluator/src/ssa/opt/constant_folding/mod.rs

```diff
@@ -276,14 +338,27 @@ impl Context {
                     }
 
                     self.values_to_replace.batch_insert(&old_results, cached);
+
+                    // Because we removed some instruction from this block, we will have to (re)visit its successors.
+                    if origin != block {
+                        origins.reused.insert(origin);
```

Comment:
I think these cascading optimizations only occur in the `NeedToHoistToCommonBlock` case. Looking at https://github.com/noir-lang/noir/issues/9963#issuecomment-3325625432, `b1` does not dominate `b3` (as every path from the entry to `b3` does not have to go through `b1`). If `b1` were to dominate `b3` we would simply remove the instruction in `b3` and be left with the single instruction in `b1`. The issue is that upon hoisting we are not actually folding the instruction that triggered the hoisting in the first place.

Reply:
You are right, and the revisits are only triggered by hoisting. This instruction is here for a different reason, it is to prevent the "missing instruction" failure in `avoid_unmapped_instructions_during_revisit`. 

There is an explanation in that test. The problem is that if between `b1` and `b3` we were to find a block `b2` which `b1` dominates, then in `b2` we would just reuse the instruction results from `b1`. Great, but then `b3` causes the instruction that created those results to be hoisted from `b3` into `b0`, and then be replaced in `b1` by the new results of the hoisted instruction. If we don't revisit `b2`, then it will have some instruction ID in its block that doesn't exist any more, so we need to go and do the `values_to_replace` chore on it.

Reply:
We could also experiment with not even trying to start at the first `origin` of the RPO and just go through all of the origins until instructions are no longer being folded.

Reply:
> although I think it would still keep the transient instructions in the DFG

I did over simplify the explanation in that comment. Some extra logic would be needed to either keep the cache information per iteration or we would have to start at the `origin`'s predecessors.

Reply:
> I thought if there are multiple origins that need revisiting, and we need a single "start" point, it would need to be the common dominator of all of them.

You're right good point.

> Sorry, I'm not sure I follow. What's the RPO?

Reverse post order

Reply:
Found the errors, hopefully the tests will be green now 🤞 

I'm a bit surprised that it's actually the "Execution Memory" which suffered, not the "Compilation Memory". Curiously there are 2 "Execution Memory" comments, and the ones at the top with the high memory (`rollup-checkpoint-root` and `rollup-checkpoint-single-block`) don't appear in the other message, and the ones that do have no noticeable difference in memory 🤔

Reply:
It looks like the performance alert is just mislabeled. We can see the compilation memory degradation here https://github.com/noir-lang/noir/pull/10013#pullrequestreview-3272465573

Reply:
Sorry I was looking at the wrong thing 🤦. If we look at the comment history for https://github.com/noir-lang/noir/pull/10013#pullrequestreview-3272465500 we see those circuits which we alerted for in the first comment but not the later ones. Possibly a bug in the reports.

Reply:
Ah those reports got cancelled https://github.com/noir-lang/noir/actions/runs/18041209670/job/51340643425?pr=10013

Reply:
Yes, I suspected those could have been the ones to bring fresh data. 

Some tests are much slower as well, in particular `execution_success/slices` takes ages, while it's fast on `master`. I'll try to see if there is some simple limit I could add to stop it from spinning its wheels.